module/bdev/aio/bdev_aio.c

```diff
@@ -231,13 +320,136 @@ bdev_aio_writev(struct file_disk *fdisk, struct spdk_io_channel *ch,
 	return len;
 }
 
+static struct aio_request_ctx *
+create_aio_request_ctx(struct spdk_bdev_io *bdev_io, aio_request_fn fn, size_t ctx_size)
+{
+	struct aio_request_ctx *request = malloc(sizeof(struct aio_request_ctx) + ctx_size);
+
+	if (request != NULL) {
+		request->aio_task = (struct bdev_aio_task *)bdev_io->driver_ctx;
+		request->thread = spdk_get_thread();
+		request->fn = fn;
+	}
+
+	return request;
+}
+
+static void
+aio_send_range_request(struct spdk_bdev_io *bdev_io, struct file_disk *fdisk, aio_request_fn fn)
+{
+	struct aio_request_ctx *request;
+	struct aio_range_ctx *ctx;
+
+	request = create_aio_request_ctx(bdev_io, fn, sizeof(struct aio_range_ctx));
+
+	if (request == NULL) {
+		spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_NOMEM);
+		return;
+	}
+
+	ctx = (struct aio_range_ctx *)request->ctx;
+
+	ctx->fdisk = fdisk;
+
+	ctx->offset_blocks = bdev_io->u.bdev.offset_blocks;
+	ctx->num_blocks = bdev_io->u.bdev.num_blocks;
+	ctx->blocklen = bdev_io->bdev->blocklen;
+
+	if (aio_send_request(request) < 0) {
+		spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_NOMEM);
+		free(request);
+	}
+}
+
+static int
+aio_flush_request_fn(void *arg)
+{
+	struct aio_flush_ctx *ctx = arg;
+
+	SPDK_NOTICELOG("[flush] file=%s\n", ctx->fdisk->filename);
+
+	return fsync(ctx->fdisk->fd);
+}
+
 static void
-bdev_aio_flush(struct file_disk *fdisk, struct bdev_aio_task *aio_task)
+bdev_aio_flush(struct spdk_bdev_io *bdev_io)
 {
-	int rc = fsync(fdisk->fd);
+	struct aio_request_ctx *request;
+	struct aio_flush_ctx *ctx;
+
+	request = create_aio_request_ctx(bdev_io, aio_flush_request_fn, sizeof(struct aio_flush_ctx));
+
+	if (request == NULL) {
+		spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_NOMEM);
+		return;
+	}
+
+	ctx = (struct aio_flush_ctx *)request->ctx;
+
+	ctx->fdisk = bdev_io->bdev->ctxt;
 
-	spdk_bdev_io_complete(spdk_bdev_io_from_ctx(aio_task),
-			      rc == 0 ? SPDK_BDEV_IO_STATUS_SUCCESS : SPDK_BDEV_IO_STATUS_FAILED);
+	if (aio_send_request(request) < 0) {
+		spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_NOMEM);
+		free(request);
+	}
+}
+
+static int
+aio_unmap_request_fn(void *arg)
+{
+	struct aio_range_ctx *ctx = arg;
+	uint64_t range[2];
+
+	range[0] = ctx->offset_blocks * ctx->blocklen;
+	range[1] = ctx->num_blocks * ctx->blocklen;
+
+	SPDK_NOTICELOG("[unmap] file=%s range=%" PRIu64 ",%" PRIu64 "\n",
+		       ctx->fdisk->filename, range[0], range[1]);
+
+	return ctx->fdisk->unmap(ctx->fdisk->fd, range);
+}
+
+static void
+bdev_aio_unmap(struct spdk_bdev_io *bdev_io)
+{
+	struct file_disk *fdisk = bdev_io->bdev->ctxt;
+
+	if (fdisk->unmap == NULL) {
+		/* We should NEVER be called in this situation as bdev_aio_io_type_supported() should return false */
```

Comment:
If so, let's use assert here

Reply:
Not sure if we want to exit in this case. All modules just fail the io request if it is of a type that is not supported.
The comment is meant to convey the idea that the caller *should* always check if an io request type is supported before submitting such requests. Maybe should just remove the comment for the sake of consistency as it is really just stating the obvious.